drivers/timer/nrf_rtc_timer.c

```diff
@@ -29,134 +37,207 @@ static struct k_spinlock lock;
-	 * next cycle.
-	 * Reconfigure CC to the future. If CC was set to next cycle we need to
-	 * wait for up to 15 us (half of 32 kHz interval) and clean a potential
-	 * event. After that there is no risk of unwanted event.
-	 */
-	prev_val = get_comparator();
-	event_clear();
-	set_comparator(now);
-	event_enable();
-
-	if (counter_sub(prev_val, now) == 1) {
-		k_busy_wait(15);
-		event_clear();
+	__ASSERT_NO_MSG(chan < CHAN_COUNT);
+	return nrf_rtc_event_address_get(RTC, nrf_rtc_compare_event_get(chan));
+}
+
+bool z_nrf_rtc_timer_compare_int_lock(uint32_t chan)
+{
+	__ASSERT_NO_MSG(chan && chan < CHAN_COUNT);
+
+	atomic_val_t prev = atomic_and(&int_mask, ~BIT(chan));
+
+	nrf_rtc_int_disable(RTC, RTC_CHANNEL_INT_MASK(chan));
+
+	return prev & BIT(chan);
+}
+
+void z_nrf_rtc_timer_compare_int_unlock(uint32_t chan, bool key)
+{
+	__ASSERT_NO_MSG(chan && chan < CHAN_COUNT);
+
+	if (key) {
+		atomic_or(&int_mask, BIT(chan));
+		nrf_rtc_int_enable(RTC, RTC_CHANNEL_INT_MASK(chan));
 	}
+}
 
-	/* Clear interrupt that may have fired as we were setting the
-	 * comparator.
-	 */
-	NVIC_ClearPendingIRQ(RTC_IRQn);
+uint32_t z_nrf_rtc_timer_compare_read(uint32_t chan)
+{
+	__ASSERT_NO_MSG(chan < CHAN_COUNT);
+
+	return nrf_rtc_cc_get(RTC, chan);
 }
 
-/* If alarm is next RTC cycle from now, function attempts to adjust. If
- * counter progresses during that time it means that 1 cycle elapsed and
- * interrupt is set pending.
- */
-static void handle_next_cycle_case(uint32_t t)
+int z_nrf_rtc_timer_get_ticks(k_timeout_t t)
 {
-	set_comparator(t + 2);
-	while (t != counter()) {
-		/* Already expired, time elapsed but event might not be
-		 * generated. Trigger interrupt.
-		 */
-		t = counter();
-		set_comparator(t + 2);
+	uint32_t curr_count;
+	int64_t curr_tick;
+	int64_t result;
+	int64_t abs_ticks;
+
+	do {
+		curr_count = counter();
+		curr_tick = z_tick_get();
+	} while (curr_count != counter());
+
+	abs_ticks = Z_TICK_ABS(t.ticks);
+	if (abs_ticks < 0) {
+		/* relative timeout */
+		return (t.ticks > COUNTER_HALF_SPAN) ?
+			-EINVAL : ((curr_count + t.ticks) & COUNTER_MAX);
 	}
+
+	/* absolute timeout */
+	result = abs_ticks - curr_tick;
+	if ((result > COUNTER_HALF_SPAN) &&
+		(result < -COUNTER_HALF_SPAN)) {
+		return -EINVAL;
+	}
+
+	return (curr_count + result) & COUNTER_MAX;
 }
 
 /* Function safely sets absolute alarm. It assumes that provided value is
- * less than MAX_CYCLES from now. It detects late setting and also handles
- * +1 cycle case.
+ * less than COUNTER_HALF_SPAN from now. It detects late setting and also
+ * handle +1 cycle case.
  */
-static void set_absolute_alarm(uint32_t abs_val)
+static void set_absolute_alarm(uint32_t chan, uint32_t abs_val)
 {
-	uint32_t diff;
-	uint32_t t = counter();
+	uint32_t now;
+	uint32_t now2;
+	uint32_t cc_val = abs_val & 0x00FFFFFF;
+	uint32_t prev_cc = get_comparator(chan);
 
-	diff = counter_sub(abs_val, t);
-	if (diff == 1) {
-		handle_next_cycle_case(t);
-		return;
-	}
+	event_enable(chan);
```

Comment:
Why the event is now enabled before the comparator is set?

Reply:
yes, it actually can be enabled later, before setting cc